### PR TITLE
New NumPyMatrix which uses NumPy methods

### PR DESCRIPTION
**Pull request by tom-pytel**
#### References to other Issues or PRs
Implements NumPyMatrix from #18253

#### Brief description of what is fixed or changed
This is basically a new matrix class which either wraps an existing NumPy matrix or creates a new one and uses it for storage and delegates some operations to NumPy for acceleration. This will deprecate the need to use `_MatrixWrapper` for NumPy matrices.

#### Release Notes

* matrices
  * Added custom NumPyMatrix which uses NumPy methods to accelerate operations.

### CODE REVIEW

**Comment by Pristine-Cat** [comment]
I've added the properties `is_mutable` and `is_sparse` because since I plan to also add an `MpMatrix` and `SciPySparseMatrix` in the future it may be desirable to have this information available in a clearer way than examining the matrix class hierarchy.

I've also thought about an `is_numeric` or `is_symbolic` property to specify whether the matrix supports abstract SymPy expressions or only numbers. Thoughts?

**Comment by Pristine-Cat** [comment]
These doctest errors seem to be because the doctest environment doesn't have numpy?

**Comment by oscarbenjamin** [comment]
> These doctest errors seem to be because the doctest environment doesn't have numpy?

Use the `doctest_depends_on` decorator.

**Comment by Pristine-Cat** [comment]
Hmm, these are very strange errors...

**Comment by sylee957** [comment]
Most of the numeric eigenvalue finding algorithms are not exact and you cannot be able to determine the multiplicity in any reasonable sense due to the nature.

Maybe `multiple=True` should be made illegal
And for the eigenvector API, the design sympy was using may not be reasonable

**Comment by Pristine-Cat** [comment]
> Most of the numeric eigenvalue finding algorithms are not exact and you cannot be able to determine the multiplicity in any reasonable sense due to the nature.

Yes, I am fighting with that right now. The NumPy `eig` function accelerates some SymPy functions which use `eigenvects` but at this moment I am fighting with a matrix where the numerical inaccuracy is giving big problems:
```
NumPyMatrix([
[ 1,  3, 0,  2],
[-2, -6, 0, -2],
[ 3,  9, 0,  0],
[-1, -3, 0,  1]]).eigenvects()
```
Compare that with:
```
Matrix([
[ 1,  3, 0,  2],
[-2, -6, 0, -2],
[ 3,  9, 0,  0],
[-1, -3, 0,  1]]).eigenvects()
```
One element of one vector is completely wrong. I can truncate the eigenvalues from `eigvals` but when I use `eig` for the vectors NumPy already calculates one vector very different from what it should be even if I correct the eigenvalues after because then it is too late.

But the errors I am talking about here are different, it succeeds the normal doctest but the TEST_ASCII="true" fails and I don't understand why, its like the `nsimplify(x, rational=True)` is not doing its job in that environment.

**Comment by oscarbenjamin** [comment]
> it succeeds the normal doctest but the TEST_ASCII="true" fails

Those are the optional dependency tests. That means that the tests fail when numpy is installed.

**Comment by Pristine-Cat** [comment]
> Those are the optional dependency tests. That means that the tests fail when numpy is installed.

But they succeed in the normal doctest with numpy installed, that is what I don't understand.

**Comment by moorepants** [comment]
I looked briefly at this and my first thought is that this doesn't follow traditional SymPy design. If you need to calculate eigenvalues using floating point operations and numeric code derived from NumPy the design pattern that SymPy has traditionally used might look something like (psuedo-code with non-existing features):

```python
m = sympy.Matrix(*symbols)
e = sympy.Eigenvalues(m)  # unevaluated (non-existing) symbolic eigenvalue operator
f = sympy.lambdify(e)
numeric_eigenvals = f(*numerical_values)
```

lambdify, autowrap, and similar tools provide a way to generate a function that does appropriate floating point operations based on the symbolic description of some set of operations.

I am not sure mixing numpy and sympy as done in this PR is something we'd want to do. Can you provide more rational and examples of why this is needed and useful?

**Comment by Pristine-Cat** [comment]
> I am not sure mixing numpy and sympy as done in this PR is something we'd want to do. Can you provide more rational and examples of why this is needed and useful?

There are two reasons for this class:

The first one is to replace `_MatrixWrapper` for numpy matrices. `_MatrixWrapper` is extremely basic and doesn't allow sympy operations to work on numpy matrices. This class will allow all sympy operations like `diagonalize`, `jordan_form`, `nullspace`, etc... to work on numpy matrices.

The second is for directly using numpy to accelerate certain sympy operations for numerical matrices. `eigenvals` and `eigenvects` are an example. I am still fighting with `eigenvects` and am not sure numpy numerical evaluation is right for this but `eigenvals` works with truncating the numerical results and rationalizing. `det` and `rank` are already accelerated and work well. Other candidates from `numpy.linalg` are `cholesky`, `qr`, `svd`, `norm`, `cond`, `trace`, `solve`, `inv`, and `pinv` - assuming numerical inaccuracy can be controlled. This is of course apart from basic operations like `__mul__` for two `NumPyMatrix` and `__pow__`. Though whenever a binary operation would take place between a numpy and a sympy matrix the sympy operation would have priority with a sympy matrix as the result.

The idea being if this works well then classes for `MpMatrix` and `SciPySparseMatrix` can also be done. This class does not require any changes to the sympy codebase itself, it just subclasses from `MutableDenseMatrix` so it will not affect any existing sympy code.

Hope that was a good rationalization...

**Comment by moorepants** [comment]
> This class will allow all sympy operations like diagonalize, jordan_form, nullspace, etc... to work on numpy matrices.

I'm not sure why that is needed. I think SymPy operators don't work on NumPy types on purpose.

> The second is for directly using numpy to accelerate certain sympy operations for numerical matrices.

This seems odd to me. SymPy matrices are not generally populated by floating point typed objects. If they happen to be, shouldn't you just call `numpy.linalg.eig(sympy.matrix2numpy(M)).` The design pattern I mention above is how one would typically move from SymPy to floating point numerical code.

**Comment by Pristine-Cat** [comment]
> I'm not sure why that is needed. I think SymPy operators don't work on NumPy types on purpose.

Actually I don't think that is the case, I recently worked on `_MatrixWrapper` (which gave me the idea to do this class) and it had a construct like this (where `self.mat` is the numpy ndarray):
```
    def __getattr__(self, attr):
        """Most attribute access is passed straight through
        to the stored matrix"""
        return getattr(self.mat, attr)
```
And this was not for `.shape` or `.rows` or `.cols` since those were already stored in the wrapper, which means they were passing on calls to the underlying numpy object for evaluation, even though those calls might have different names or return non-sympified objects. The intent seems to have been to access the underlying numpy methods. It seems a weird idea to think that SymPy is actually less compatible with NumPy on purpose? Why do that?

> This seems odd to me. SymPy matrices are not generally populated by floating point typed objects. If they happen to be, shouldn't you just call `numpy.linalg.eig(sympy.matrix2numpy(M)).` The design pattern I mention above is how one would typically move from SymPy to floating point numerical code.

You will then need to do some or even a good deal of cleanup due to numpy numerical inaccuracies, which this class does for you. As for pure numerical matrices in sympy, they have some problems of their own, for example:
```
Matrix(4, 4, [0, 1, 2, 3, 4, 5, 6, 7, 2, 9, 10, 11, 12, 13, 14, 14]).eigenvects()
```
That never terminates, whereas:
```
NumPyMatrix(4, 4, [0, 1, 2, 3, 4, 5, 6, 7, 2, 9, 10, 11, 12, 13, 14, 14]).eigenvects()
```
is instant (if not yet completely resolved correctly). Or for example:
```
M = Matrix([[S('1+I'),S('-19/4+5/4*I'),S('1/2-I'),S('9/4+55/16*I'),S('-3/4'),S('45/32-37/16*I'),S('1/4+1/2*I'),S('-129/64-9/64*I'),S('1/4-5/16*I'),S('65/128+87/64*I'),S('-9/32-1/16*I'),S('183/256-97/128*I'),S('3/64+13/64*I'),S('-23/32-59/256*I'),S('15/128-3/32*I'),S('19/256+551/1024*I')],
        [S('21/8+I'),S('-537/64+143/16*I'),S('-5/8-39/16*I'),S('2473/256+137/64*I'),S('-149/64+49/32*I'),S('-177/128-1369/128*I'),S('125/64+87/64*I'),S('-2063/256+541/128*I'),S('85/256-33/16*I'),S('805/128+2415/512*I'),S('-219/128+115/256*I'),S('6301/4096-6609/1024*I'),S('119/128+143/128*I'),S('-10879/2048+4343/4096*I'),S('129/256-549/512*I'),S('42533/16384+29103/8192*I')],
        [S('-2'),S('17/4-13/2*I'),S('1+I'),S('-19/4+5/4*I'),S('1/2-I'),S('9/4+55/16*I'),S('-3/4'),S('45/32-37/16*I'),S('1/4+1/2*I'),S('-129/64-9/64*I'),S('1/4-5/16*I'),S('65/128+87/64*I'),S('-9/32-1/16*I'),S('183/256-97/128*I'),S('3/64+13/64*I'),S('-23/32-59/256*I')],
        [S('1/4+13/4*I'),S('-825/64-147/32*I'),S('21/8+I'),S('-537/64+143/16*I'),S('-5/8-39/16*I'),S('2473/256+137/64*I'),S('-149/64+49/32*I'),S('-177/128-1369/128*I'),S('125/64+87/64*I'),S('-2063/256+541/128*I'),S('85/256-33/16*I'),S('805/128+2415/512*I'),S('-219/128+115/256*I'),S('6301/4096-6609/1024*I'),S('119/128+143/128*I'),S('-10879/2048+4343/4096*I')],
        [S('-4*I'),S('27/2+6*I'),S('-2'),S('17/4-13/2*I'),S('1+I'),S('-19/4+5/4*I'),S('1/2-I'),S('9/4+55/16*I'),S('-3/4'),S('45/32-37/16*I'),S('1/4+1/2*I'),S('-129/64-9/64*I'),S('1/4-5/16*I'),S('65/128+87/64*I'),S('-9/32-1/16*I'),S('183/256-97/128*I')],
        [S('1/4+5/2*I'),S('-23/8-57/16*I'),S('1/4+13/4*I'),S('-825/64-147/32*I'),S('21/8+I'),S('-537/64+143/16*I'),S('-5/8-39/16*I'),S('2473/256+137/64*I'),S('-149/64+49/32*I'),S('-177/128-1369/128*I'),S('125/64+87/64*I'),S('-2063/256+541/128*I'),S('85/256-33/16*I'),S('805/128+2415/512*I'),S('-219/128+115/256*I'),S('6301/4096-6609/1024*I')],
        [S('-4'),S('9-5*I'),S('-4*I'),S('27/2+6*I'),S('-2'),S('17/4-13/2*I'),S('1+I'),S('-19/4+5/4*I'),S('1/2-I'),S('9/4+55/16*I'),S('-3/4'),S('45/32-37/16*I'),S('1/4+1/2*I'),S('-129/64-9/64*I'),S('1/4-5/16*I'),S('65/128+87/64*I')],
        [S('-2*I'),S('119/8+29/4*I'),S('1/4+5/2*I'),S('-23/8-57/16*I'),S('1/4+13/4*I'),S('-825/64-147/32*I'),S('21/8+I'),S('-537/64+143/16*I'),S('-5/8-39/16*I'),S('2473/256+137/64*I'),S('-149/64+49/32*I'),S('-177/128-1369/128*I'),S('125/64+87/64*I'),S('-2063/256+541/128*I'),S('85/256-33/16*I'),S('805/128+2415/512*I')],
        [S('0'),S('-6'),S('-4'),S('9-5*I'),S('-4*I'),S('27/2+6*I'),S('-2'),S('17/4-13/2*I'),S('1+I'),S('-19/4+5/4*I'),S('1/2-I'),S('9/4+55/16*I'),S('-3/4'),S('45/32-37/16*I'),S('1/4+1/2*I'),S('-129/64-9/64*I')],
        [S('1'),S('-9/4+3*I'),S('-2*I'),S('119/8+29/4*I'),S('1/4+5/2*I'),S('-23/8-57/16*I'),S('1/4+13/4*I'),S('-825/64-147/32*I'),S('21/8+I'),S('-537/64+143/16*I'),S('-5/8-39/16*I'),S('2473/256+137/64*I'),S('-149/64+49/32*I'),S('-177/128-1369/128*I'),S('125/64+87/64*I'),S('-2063/256+541/128*I')],
        [S('0'),S('-4*I'),S('0'),S('-6'),S('-4'),S('9-5*I'),S('-4*I'),S('27/2+6*I'),S('-2'),S('17/4-13/2*I'),S('1+I'),S('-19/4+5/4*I'),S('1/2-I'),S('9/4+55/16*I'),S('-3/4'),S('45/32-37/16*I')],
        [S('0'),S('1/4+1/2*I'),S('1'),S('-9/4+3*I'),S('-2*I'),S('119/8+29/4*I'),S('1/4+5/2*I'),S('-23/8-57/16*I'),S('1/4+13/4*I'),S('-825/64-147/32*I'),S('21/8+I'),S('-537/64+143/16*I'),S('-5/8-39/16*I'),S('2473/256+137/64*I'),S('-149/64+49/32*I'),S('-177/128-1369/128*I')]])
M.rank()
```
That also never finishes where:
```
NumPyMatrix(M).rank()
```
is instant. In fact anything in sympy which uses eigenvects tends to die fairly quickly for larger matrices which is why I am trying so hard to get that working well since it is a basis for many other operations.

**Comment by sylee957** [comment]
```
M = Matrix([[S('1+I'),S('-19/4+5/4*I'),S('1/2-I'),S('9/4+55/16*I'),S('-3/4'),S('45/32-37/16*I'),S('1/4+1/2*I'),S('-129/64-9/64*I'),S('1/4-5/16*I'),S('65/128+87/64*I'),S('-9/32-1/16*I'),S('183/256-97/128*I'),S('3/64+13/64*I'),S('-23/32-59/256*I'),S('15/128-3/32*I'),S('19/256+551/1024*I')],
        [S('21/8+I'),S('-537/64+143/16*I'),S('-5/8-39/16*I'),S('2473/256+137/64*I'),S('-149/64+49/32*I'),S('-177/128-1369/128*I'),S('125/64+87/64*I'),S('-2063/256+541/128*I'),S('85/256-33/16*I'),S('805/128+2415/512*I'),S('-219/128+115/256*I'),S('6301/4096-6609/1024*I'),S('119/128+143/128*I'),S('-10879/2048+4343/4096*I'),S('129/256-549/512*I'),S('42533/16384+29103/8192*I')],
        [S('-2'),S('17/4-13/2*I'),S('1+I'),S('-19/4+5/4*I'),S('1/2-I'),S('9/4+55/16*I'),S('-3/4'),S('45/32-37/16*I'),S('1/4+1/2*I'),S('-129/64-9/64*I'),S('1/4-5/16*I'),S('65/128+87/64*I'),S('-9/32-1/16*I'),S('183/256-97/128*I'),S('3/64+13/64*I'),S('-23/32-59/256*I')],
        [S('1/4+13/4*I'),S('-825/64-147/32*I'),S('21/8+I'),S('-537/64+143/16*I'),S('-5/8-39/16*I'),S('2473/256+137/64*I'),S('-149/64+49/32*I'),S('-177/128-1369/128*I'),S('125/64+87/64*I'),S('-2063/256+541/128*I'),S('85/256-33/16*I'),S('805/128+2415/512*I'),S('-219/128+115/256*I'),S('6301/4096-6609/1024*I'),S('119/128+143/128*I'),S('-10879/2048+4343/4096*I')],
        [S('-4*I'),S('27/2+6*I'),S('-2'),S('17/4-13/2*I'),S('1+I'),S('-19/4+5/4*I'),S('1/2-I'),S('9/4+55/16*I'),S('-3/4'),S('45/32-37/16*I'),S('1/4+1/2*I'),S('-129/64-9/64*I'),S('1/4-5/16*I'),S('65/128+87/64*I'),S('-9/32-1/16*I'),S('183/256-97/128*I')],
        [S('1/4+5/2*I'),S('-23/8-57/16*I'),S('1/4+13/4*I'),S('-825/64-147/32*I'),S('21/8+I'),S('-537/64+143/16*I'),S('-5/8-39/16*I'),S('2473/256+137/64*I'),S('-149/64+49/32*I'),S('-177/128-1369/128*I'),S('125/64+87/64*I'),S('-2063/256+541/128*I'),S('85/256-33/16*I'),S('805/128+2415/512*I'),S('-219/128+115/256*I'),S('6301/4096-6609/1024*I')],
        [S('-4'),S('9-5*I'),S('-4*I'),S('27/2+6*I'),S('-2'),S('17/4-13/2*I'),S('1+I'),S('-19/4+5/4*I'),S('1/2-I'),S('9/4+55/16*I'),S('-3/4'),S('45/32-37/16*I'),S('1/4+1/2*I'),S('-129/64-9/64*I'),S('1/4-5/16*I'),S('65/128+87/64*I')],
        [S('-2*I'),S('119/8+29/4*I'),S('1/4+5/2*I'),S('-23/8-57/16*I'),S('1/4+13/4*I'),S('-825/64-147/32*I'),S('21/8+I'),S('-537/64+143/16*I'),S('-5/8-39/16*I'),S('2473/256+137/64*I'),S('-149/64+49/32*I'),S('-177/128-1369/128*I'),S('125/64+87/64*I'),S('-2063/256+541/128*I'),S('85/256-33/16*I'),S('805/128+2415/512*I')],
        [S('0'),S('-6'),S('-4'),S('9-5*I'),S('-4*I'),S('27/2+6*I'),S('-2'),S('17/4-13/2*I'),S('1+I'),S('-19/4+5/4*I'),S('1/2-I'),S('9/4+55/16*I'),S('-3/4'),S('45/32-37/16*I'),S('1/4+1/2*I'),S('-129/64-9/64*I')],
        [S('1'),S('-9/4+3*I'),S('-2*I'),S('119/8+29/4*I'),S('1/4+5/2*I'),S('-23/8-57/16*I'),S('1/4+13/4*I'),S('-825/64-147/32*I'),S('21/8+I'),S('-537/64+143/16*I'),S('-5/8-39/16*I'),S('2473/256+137/64*I'),S('-149/64+49/32*I'),S('-177/128-1369/128*I'),S('125/64+87/64*I'),S('-2063/256+541/128*I')],
        [S('0'),S('-4*I'),S('0'),S('-6'),S('-4'),S('9-5*I'),S('-4*I'),S('27/2+6*I'),S('-2'),S('17/4-13/2*I'),S('1+I'),S('-19/4+5/4*I'),S('1/2-I'),S('9/4+55/16*I'),S('-3/4'),S('45/32-37/16*I')],
        [S('0'),S('1/4+1/2*I'),S('1'),S('-9/4+3*I'),S('-2*I'),S('119/8+29/4*I'),S('1/4+5/2*I'),S('-23/8-57/16*I'),S('1/4+13/4*I'),S('-825/64-147/32*I'),S('21/8+I'),S('-537/64+143/16*I'),S('-5/8-39/16*I'),S('2473/256+137/64*I'),S('-149/64+49/32*I'),S('-177/128-1369/128*I')]])
M.rank()
```

Actually, `NumPy` is not a perfect way to solve this problem. For example, there are a lot of numerical precision loss and ill-conditioning problem if you use the floating points. 
For numeric workflow, there are plenty of materials regarding such precision issues making wrong results for computing matrix RREF or Jordan decomposition.

I would say that the 'best' way to resolve this sort of problem in symbolic library is introducing the gaussian rational, or even further a generalized 'algebraic number field', as far as I have studied for.

Although I don't think that the above comment would be faster than numpy, but it should be more faster than what we already have now, in theoretical sense.

**Comment by oscarbenjamin** [comment]
> ```
> M.rank()
> ```
> That also never finishes

This is reasonably fast :)
```
In [3]: M.rank(dotprodsimp=True)                                                                                                               
Out[3]: 8
```

**Comment by Pristine-Cat** [comment]
> This is reasonably fast :)
> 
> ```
> In [3]: M.rank(dotprodsimp=True)                                                                                                               
> Out[3]: 8
> ```

Yes I know, I was just trying to win the argument :)

**Comment by Pristine-Cat** [comment]
> Actually, `NumPy` is not a perfect way to solve this problem. For example, there are a lot of numerical precision loss and ill-conditioning problem if you use the floating points.
> There are plenty of materials regarding the fact how such precision issues can make wrong results for computing matrix RREF or Jordan decomposition.

I am leaning to let numpy calculate the eigenvals since those can be corrected but go to sympy for the eigenvects.

**Comment by moorepants** [comment]
It is hard for me to see why we'd ever want NumPy to calculate anything in a SymPy object. SymPy is a library built to operate on symbolic and arbitrary precision types. It is somewhat conicendental that it can operate on floating point types too. Yes, we want eigenvalue and eigenvector calculations to be correct (and fast), but for symbolic and arbitrary precision types. If you are using SymPy objects as containers for floating point types you are likely using the wrong library.

File: sympy/matrices/numpy.py [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+        else:
+            if any(isinstance(e, Float) for e in l) or i != f:
+                return np.float64
+            else:
+                return np.int64
+
+    try:
+        all(complex(e) for e in l)
+    except TypeError: # non-numeric source data
+        raise TypeError('source data for NumPyMatrix must be numeric') from None
```
Comment (czgdp1807) [inline sympy/matrices/numpy.py]: Can we do the same with `if-else`?

**Comment by Pristine-Cat** [comment]
> It is hard for me to see why we'd ever want NumPy to calculate anything in a SymPy object. SymPy is a library built to operate on symbolic and arbitrary precision types. It is somewhat conicendental that it can operate on floating point types too. Yes, we want eigenvalue and eigenvector calculations to be correct (and fast), but for symbolic and arbitrary precision types. If you are using SymPy objects as containers for floating point types you are likely using the wrong library.

Yes, I think I am giving up on the floating point accelerations, too inaccurate and unpredictable. Its just very frustrating when I give SymPy what should be a very easy problem and it hangs. There are other cases though like certain operations on integer matrices which would not suffer from floating point inaccuracies, for example - `randMatrix(50).rank()` hangs while `NumPyMatrix(randMatrix(50)).rank()` gives (and should always give) the expected result. The determinant of an integer matrix is always an integer so anything based on that could see a benefit.

In general, even if I do not wind up accelerating the vast majority of cases then this still winds up as a `_MatrixWrapper` which actually works.

File: sympy/matrices/numpy.py (hunk 1 above)
Comment (Pristine-Cat) [reply sympy/matrices/numpy.py]: Please elaborate... (Keep in mind that at this point the objects in the list could be anything, Python, SymPy, mpf objects, anything really it comes from the user).

**Comment by Pristine-Cat** [comment]
> I would say that the 'best' way to resolve this sort of problem in symbolic library is introducing the gaussian rational, or even further, a generalized 'algebraic number field', as far as I have found the answers in the books.

By gaussian rational you mean a `Complex` class which essentially encapsulates the real and `I` component as rationals?

**Comment by sylee957** [comment]
> > I would say that the 'best' way to resolve this sort of problem in symbolic library is introducing the gaussian rational, or even further, a generalized 'algebraic number field', as far as I have found the answers in the books.
> 
> By gaussian rational you mean a `Complex` class which essentially encapsulates the real and `I` component as rationals?

Yes, it is work in progress like #15396

**Comment by Pristine-Cat** [comment]
Well, I've added basic operations and `_matrixify` to this instead of the old wrapper. While float matrices might be anathema to the SymPy community (especially limited native precision) they are used out in the wild, though whether that is something for a symbolic library is a valid question. Where this can potentially help however without loss of precision apart from being just a wrapper for `ndarray` is for integer matrices - granted limited to 64 bits - there are use cases for those.

Compare:
```
$ python -m timeit -s "import sympy as sp; m = sp.randMatrix(100)" "m**2"
10 loops, best of 3: 5.84 sec per loop
```
to:
```
$ python -m timeit -s "import sympy as sp; m = sp.NumPyMatrix(sp.randMatrix(100))" "m**2"
100 loops, best of 3: 2.89 msec per loop
```
`_op_priority` and `_class_priority` are set so that SymPy operations and classes have precedence, so if you mix one of these in a binary op with a native SymPy matrix then the native SymPy operation prevails and returns a native SymPy matrix. NumPy basic operations are used when operating on two `NumPyMatrix` objects or a `NumPyMatrix` and a scalar. Non-scalars result in SymPy matrices.

So what do you guys think?

**Comment by Pristine-Cat** [comment]
Hmm, PY2 error, I thought PY2 was deprecated?

**Comment by sylee957** [comment]
It's not possible to use machine integer matrix to replace the sympy feature perfectly, because of the overflow issues.
The only place where numpy integer matrix can beat stuff in symbolic math is for 'modular integer' computations, because of the nature that it only needs a fixed amount of storage.

**Comment by asmeurer** [comment]
I don't understand the point of this. Sorry if this was already answered above, but why would you use this over just `numpy.array` with `numpy.linalg` and `scipy.linalg` functions? Are there SymPy matrix algorithms that aren't implemented for numpy arrays that you would want to apply to a numeric array?

**Comment by Pristine-Cat** [comment]
Well it was a shot at putting the SymPy API on underlying NumPy matrices for convenience as well as accelerating some operations. Though the accelerations for floating point are not worth it for the inaccuracies and it doesn't in general fit with the symbolic precise nature of SymPy so I will call it a failed experiment since it doesn't really solve a big underlying problem anyway. Had fun writing it though and learned some more about matrices :)

**Comment by asmeurer** [comment]
If you want to write code that works uniformly for both SymPy matrices and numpy matrices, I would suggest using matrix expressions with lambdify (some things may not work yet, but they should be fixed).

**Pull request closed without merging**